src/System.Text.Json/src/System/Text/Json/Serialization/JsonPropertyInfo.cs

```diff
@@ -345,29 +361,62 @@ public void Read(JsonTokenType tokenType, ref ReadStack state, ref Utf8JsonReade
                 // Forward the setter to the value-based JsonPropertyInfo.
                 propertyInfo.ReadEnumerable(tokenType, ref state, ref reader);
             }
+            // For performance on release build, don't verify converter correctness for internal converters.
+            else if (HasInternalConverter)
```

Comment:
Is this special-casing/complexity worth the perf gain? I don't know if we should treat the internal converters this way. We are saving a couple of branches and introducing 1. If this change doesn't have a big difference, I would revert it.

Reply:
I ran the deserialize benchmark 4 times over each of the following factoring and picked the fastest time (the `Min` column) in the bechmark. Doing that seems to produce the most accurate result. Here's the data:
- Current code in PR: 1331us
- Previous code: 1377us
- Current code + `AggressiveInlining` on `VerifyRead`: 1361ns

so (1361-1331) / 1361 = ~2.2% faster.

I propose to keep the changes in the PR since ~2% is significant and because it affects every property read (or write).

Reply:
> Doing that seems to produce the most accurate result

@adamsitnik - is that the approach you suggest we follow?

I would have though that a big value-prop of using BDN was that the "Median"/actual result is what we would use for perf comparison and the error column quantifies the variation. If the variation/error is ~2%, then I wouldn't take the results you shared as a strong indication of a perf improvement. However, I don't think the error is that high at all (from the results you shared). Why not use the "Median" column?

> I propose to keep the changes in the PR since ~2% is significant and because it affects every property read (or write).

Fair enough. Sounds good to me.

Reply:
Note the profiler also shows ~2% for OnRead() which is mostly composed of calling VerifyRead() and its overhead of gathering parameters.

Using "Median" on my local machine is OK but doesn't give me a best-case "non-interference" value that I need for very small gains that are close to the margin of error. I may have apps open including VS which will be doing background compilation etc. A small burp shifts the median. Also by using "Min" a few times I can see if there was random interference and I should run again or close some offending apps (such as a browser or VS).

Reply:
> is that the approach you suggest we follow?

I would recommend comparing the entire distribution. Performance is never a single value. "Statistics for Performance Engineers" from [Pro .NET Benchmarking](https://aakinshin.net/prodotnetbenchmarking/) contains a great answer to this question.

When it comes to getting super stable results for validating 2% gain you can tell BDN to affinitize the benchmark process to a given CPU. It works great for single-threaded benchmarks. 

```cmd
--affinity $value
```
`$value` is an integer representing the bit mask send to [OS API](
https://msdn.microsoft.com/library/system.diagnostics.process.processoraffinity.aspx) so for example if you send 8 it's translated to `1000` which affinitizes the benchmark to 4th Processor

btw you can also tell it to run the benchmark processes many times

```cmd
--launchCount $times
```

Reply:
> Is this special-casing/complexity worth the perf gain?

BTW @sebastienros is going to have a modern BCL talk on 1st on November when he is going to show us how to use ASP.NET Perf lab to run TechEmpower benchmarks using local builds of CoreFX to validate if the change is worth it from TechEmpower scenario perspective. I personally use it anytime I am not sure if a small gain in microbenchmarks is worth the extra complexity.